arrow.qmd

```diff
@@ -0,0 +1,292 @@
+
+### Prerequisites
+
+In this chapter, we'll continue to use the tidyverse, particularly dplyr, but now we'll pair it with the arrow package which is designed specifically for working with large data.
+
+```{r setup}
+#| message: false
+#| warning: false
+library(tidyverse)
+library(arrow)
+```
+
+Later in the chapter, we'll also see some connections between arrow and duckdb, so we'll also need dbplyr and duckdb.
+
+```{r}
+library(dbplyr)
+library(duckdb)
+```
+
+## Getting the data
+
+We being by getting a dataset worthy of these tools: a data set of item checkouts from Seattle public libraries, available online at [data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6](https://data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6).
+This dataset contains 41,389,465 rows of describing for each book, for each type (e.g. book, ebook, CD), how many times it was checked out each month.
+
+The data is distributed as a single 9GB CSV file, so it may take some time to download: this is one of the challenges of working with bigger datasets!
+The following code will get you a cached copy of the data:
+
+```{r}
+#| eval: false
+dir.create("data", showWarnings = FALSE)
+url <- "https://r4ds.s3.us-west-2.amazonaws.com/seattle-library-checkouts.csv"
+
+# Default timeout is 60s; bump it up to an hour
+options(timeout = 60 * 60)
+download.file(url, "data/seattle-library-checkouts.csv")
+```
+
+## Opening a dataset
+
+Let's start by taking a look at the Seattle library checkout data.
+At 9GB, this file is large enough that we probably don't want to load the whole thing into memory.
+A good rule of thumb is that you usually want at least twice as much memory as the size of the data, and many laptops top out at 16 Gb.
+So we want to avoid `read_csv()`, and instead use the `open_dataset()` function from the arrow package:
+
+```{r open-dataset}
+# partial schema for ISBN column only
+opts <- CsvConvertOptions$create(col_types = schema(ISBN = string()))
+
+seattle_csv <- open_dataset(
+  sources = "data/seattle-library-checkouts.csv", 
+  format = "csv",
+  convert_options = opts
+)
+```
+
+(Here we've had to use some relatively advanced code to parse the ISBN variable correctly: this is because the first \~83,000 rows don't contain any data so arrow guesses the wrong types. The arrow team is aware of this problem and there will hopefully be a better approach by the time you read this chapter.)
+
+What happens when this code is run?
+`open_dataset()` will scan a few thousand rows to figure out the structure of the data set.
+Then it records what it's found and stops; it will only read further rows as you specifically request them.
+This metadata is what we see if we print `seattle_csv`:
+
+```{r}
+seattle_csv
+```
+
+The first line in the output tell you that `seattle_csv` is stored locally on-disk as a single csv file; it will only be loaded into memory when necessary.
```

Comment:
```suggestion
The first line in the output tells you that `seattle_csv` is stored locally on-disk as a single CSV file; it will only be loaded into memory when necessary.
```